Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,400 @@
+            }
+            factory.diagnosticsHandler(diagnostic)
+        }
+    }
+
+    // FIXME: we want to remove this functionality and move to more conventional error handling
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
+    public var errorsReported: Bool {
+        self._errorsReported.get() ?? false
+    }
+
+    private struct NOOPFactory: ObservabilityFactory {
+        var diagnosticsHandler: DiagnosticsHandler = { _ in }
+    }
+}
+
+public protocol ObservabilityFactory {
+    var diagnosticsHandler: DiagnosticsHandler { get }
+}
+
+public typealias DiagnosticsHandler = (Diagnostic) -> Void
+
+public struct Diagnostic: CustomStringConvertible, Equatable {
+    public let severity: Severity
+    public let message: String
+    public internal (set) var metadata: DiagnosticsMetadata?
+
+    public init(severity: Severity, message: String, metadata: DiagnosticsMetadata?) {
+        self.severity = severity
+        self.message = message
+        self.metadata = metadata
+    }
+
+    public var description: String {
+        return "[\(self.severity)]: \(self.message)"
+    }
+
+    public static func == (lhs: Diagnostic, rhs: Diagnostic) -> Bool {
+        if lhs.severity != rhs.severity {
+            return false
+        }
+        if lhs.message != rhs.message {
+            return false
+        }
+        // FIXME
+        /*
+         if lhs.metadata != rhs.metadata {
+         return false
+         }*/
+        return true
+    }
+
+    public static func error(_ message: String, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .error, message: message, metadata: metadata)
+    }
+
+    public static func error(_ message: CustomStringConvertible, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .error, message: message.description, metadata: metadata)
+    }
+
+    public static func warning(_ message: String, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .warning, message: message, metadata: metadata)
+    }
+
+    public static func warning(_ message: CustomStringConvertible, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .warning, message: message.description, metadata: metadata)
+    }
+
+    public static func info(_ message: String, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .info, message: message, metadata: metadata)
+    }
+
+    public static func info(_ message: CustomStringConvertible, metadata: DiagnosticsMetadata? = .none) -> Self {
+        Self(severity: .info, message: message.description, metadata: metadata)
+    }
+
+    public enum Severity: Equatable {
+        case error
+        case warning
+        case info
+        case debug
+    }
+}
+
+// TODO: consider using @autoclosure to delay potentially expensive evaluation of data when some diagnostics may be filtered out
```

Comment:
is the flip side is that this would capture and all state until the diagnostic was handled?  Would the ideal be to quickly make the decision and their either call the closure or discard it, to avoid lengthening the life of anything capture by the closure?